async/thread_safe_queue.hpp

```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+
+// A threadsafe-queue.
+template <class T>
+class thread_safe_queue_t {
+    std::queue<T> queue;
```

Comment:
Would suggest making the template parameter for the container type itself rather than the value type of the container:
```suggestion
template <class TContainer>
class thread_safe {
    TContainer container;
```